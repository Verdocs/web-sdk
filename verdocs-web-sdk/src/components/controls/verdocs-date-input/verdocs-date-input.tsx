import AirDatepicker from 'air-datepicker';
import {Component, Element, Prop, Host, h, State, Method} from '@stencil/core';
import {VerdocsToast} from '../../../utils/Toast';
import localeEn from 'air-datepicker/locale/en';

/**
 * Display a date input field.
 *
 * ```ts
 * <verdocs-date-input type="text" label="DOB" placeholder="Date of Birth..." value="" />
 * ```
 */
@Component({
  tag: 'verdocs-date-input',
  styleUrl: 'verdocs-date-input.scss',
})
export class VerdocsDateInput {
  private inputEl: HTMLInputElement;
  private picker: AirDatepicker<HTMLElement> | null = null;

  @Element()
  hostEl: HTMLElement;

  /**
   * The initial value for the input field.
   */
  @Prop({mutable: true, reflect: true}) value: string = '';

  /**
   * The label for the field.
   */
  @Prop() label: string = '';

  /**
   * The placeholder for the field.
   */
  @Prop() placeholder: string = '';

  /**
   * If supplied, a help icon will be displayed to provide the user more information.
   */
  @Prop() helpText: string = '';

  /**
   * Should the field be disabled?
   */
  @Prop() disabled: boolean = false;

  /**
   * Should the field be required?
   */
  @Prop() required: boolean = false;

  @State() showingPw: boolean = false;
  @State() focused?: boolean = false;
  @State() containerId = `verdocs-date-input-${Math.random().toString(36).substring(2, 11)}`;

  @Method()
  async focusField() {
    // Our input field is fake, so we fake the flash too
    this.focused = true;
    this.picker?.show();
    setTimeout(() => {
      this.focused = false;
    }, 500);
  }

  componentDidLoad() {
    this.picker = new AirDatepicker<HTMLElement>(`#${this.containerId}`, {
      locale: localeEn,
      isMobile: true,
      autoClose: true,
      onShow: () => (this.focused = true),
      onHide: () => (this.focused = false),
      onSelect: ({date, formattedDate}) => {
        const event = new CustomEvent('input', {detail: {date, formattedDate}});
        this.hostEl.dispatchEvent(event);
      },
    });
  }

  copyToClipboard() {
    this.inputEl.select();
    navigator.clipboard
      .writeText(this.value)
      .then(() => VerdocsToast('Copied!'))
      .catch(e => console.log('Error copying to clipboard', e));
  }

  render() {
    return (
      <Host class={`input-field`}>
        <label>
          {this.label && (
            <div class="input-label">
              {this.label + ':'}
              {this.required && <span class="required">*</span>}
            </div>
          )}

          <div class="input-container">
            <input
              type="text"
              value={this.value}
              required={this.required}
              class="input-element"
              data-lpignore="true"
              id={this.containerId}
              disabled={this.disabled}
              placeholder={this.placeholder}
              // onBlur={(e: any) => this.blur?.emit(e.target.value)}
              ref={el => (this.inputEl = el as HTMLInputElement)}
              onInput={(e: any) => (this.value = e.target.value)}
            />
          </div>

          {this.helpText && <verdocs-help-icon text={this.helpText} />}
        </label>
      </Host>
    );
  }
}
