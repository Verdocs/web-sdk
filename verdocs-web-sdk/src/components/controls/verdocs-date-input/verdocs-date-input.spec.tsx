import { newSpecPage } from '@stencil/core/testing';
import { VerdocsDateInput } from './verdocs-date-input';
import { h } from '@stencil/core';

describe('verdocs-date-input', () => {
  beforeAll(() => {
    jest.spyOn(Math, 'random').mockReturnValue(0.123456789);
  });

  afterAll(() => {
    // restore the original behavior after the tests
    (Math.random as jest.Mock).mockRestore();
  });
  it('renders with default props', async () => {
    const page = await newSpecPage({
      components: [VerdocsDateInput],
      html: `<verdocs-date-input></verdocs-date-input>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('renders with label, placeholder, and value', async () => {
    const page = await newSpecPage({
      components: [VerdocsDateInput],
      template: () => <verdocs-date-input label="DOB" placeholder="Date of Birth..." value="2023-01-01"></verdocs-date-input>,
    });
    expect(page.root).toMatchSnapshot();
    const input = page.root.querySelector('input');
    expect(input?.value).toBe('2023-01-01');
  });

  it('renders as disabled', async () => {
    const page = await newSpecPage({
      components: [VerdocsDateInput],
      template: () => <verdocs-date-input disabled={true}></verdocs-date-input>,
    });
    expect(page.root).toMatchSnapshot();
    const input = page.root.querySelector('input');
    expect(input?.disabled).toBe(true);
  });

  it('handles input change event', async () => {
    const page = await newSpecPage({
      components: [VerdocsDateInput],
      template: () => <verdocs-date-input></verdocs-date-input>,
    });
    const input = page.root.querySelector('input');
    expect(input).not.toBeNull();
    if (input) {
      input.value = '2025-07-23';
      input.dispatchEvent(new Event('input'));
      expect(input.value).toBe('2025-07-23');
    }
  });
});
