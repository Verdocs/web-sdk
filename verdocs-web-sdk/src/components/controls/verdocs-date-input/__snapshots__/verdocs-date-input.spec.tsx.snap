// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verdocs-date-input renders as disabled 1`] = `
<verdocs-date-input class="input-field" value="">
  <label>
    <div class="input-container">
      <input class="input-element" data-lpignore="true" disabled="" id="verdocs-date-input-4fzzzxjyl" placeholder="" readonly="true" type="text" value="">
    </div>
  </label>
</verdocs-date-input>
`;

exports[`verdocs-date-input renders with default props 1`] = `
<verdocs-date-input class="input-field" value="">
  <label>
    <div class="input-container">
      <input class="input-element" data-lpignore="true" id="verdocs-date-input-4fzzzxjyl" placeholder="" readonly="true" type="text" value="">
    </div>
  </label>
</verdocs-date-input>
`;

exports[`verdocs-date-input renders with label, placeholder, and value 1`] = `
<verdocs-date-input class="input-field" value="2023-01-01">
  <label>
    <div class="input-label">
      DOB:
    </div>
    <div class="input-container">
      <input class="input-element" data-lpignore="true" id="verdocs-date-input-4fzzzxjyl" placeholder="Date of Birth..." readonly="true" type="text" value="2023-01-01">
    </div>
  </label>
</verdocs-date-input>
`;
